### ISSUE TITLE
gltfpack produces broken result for skinned parts of mesh

### ISSUE THREAD

**Issue by hybridherbst** [zeux/meshoptimizer#433]
1. Download https://github.com/prefrontalcortex/glTF-Sample-Models/raw/carbon-frame-bike/2.0/CarbonFrameBike/glTF-Binary/CarbonFrameBike.glb
2. See reference images at https://github.com/prefrontalcortex/glTF-Sample-Models/tree/carbon-frame-bike/2.0/CarbonFrameBike
3. Run `gltfpack -i CarbonFrameBike.glb -o CarbonFrameBike.packed.glb`
4. Open the result in e.g. [Babylon Sandbox](https://sandbox.babylonjs.com/) or Gestaltor
5. Note that the wires don't survive packing well:
  ![image](https://user-images.githubusercontent.com/2693840/178118767-575834d6-8445-408a-8025-00618e8fcf6c.png)

Reference: 
![image](https://user-images.githubusercontent.com/2693840/178118964-dd1a972b-553c-49a7-82f5-e639de0e53b7.png)

Using `-noq` works around the problem, but I thought it would be worth still raising an issue since the result looks relatively broken with default parameters.

**Comment by zeux** [comment]
fwiw this looks as if the default number of bits (14) gltfpack uses for quantization isn't sufficient for this model. In certain cases this can be avoided by recentering the mesh, which gltfpack doesn't do automatically, but often it's just something that needs to be raised via `-vp` command line flag.

I'll keep this open for a bit because I need to check if gltfpack can improve this automatically by changing the mesh parameters, but using something like `-vp 16` is likely to improve this noticeably.

**Comment by zeux** [comment]
Right, so:

1. This can be improved via `-vp 16`
2. This happens because the wires in question have a very large internal scale value (of 100 - this looks like an improper coordinate system conversion somewhere; probably the inverse scale of 1/100 was baked into the vertex data). Since quantization is done in local space, but using a single global quantization grid (relaxing this is difficult because eg several meshes can reuse skin objects and the quantization scale needs to be corrected via skinning transforms), the vertices end up being quantized with suboptimal values.

I think you don't see the issue in glTF-Transform because it's not quantizing the values, either because it just doesn't support quantization on skinned meshes, or because it requires an explicit command line instruction to enable that.

Either way, I don't think this can realistically be fixed in quantization code, and propagating the scale value "into" mesh vertices would result in different bone transforms which may be important for other applications.

**Comment by zeux** [comment]
P.S. The node with an abnormally high scale (100) that causes all this is Armature_S1/S2/S3. Maybe this is an artifact of a Blender export judging by the name, or maybe the source model is just configured in an odd way.

**Comment by zeux** [comment]
I guess the one thing that gltfpack could do is compute two quantization grids, one for skinned meshes and one for non-skinned meshes... but it's a little silly to do that as it assumes that all or none of the skinned meshes in the model are poorly quantized due to abnormal scale values. If this example comes up in the future that could be a possible option but for now it feels like this shouldn't be handled by gltfpack.

**Comment by donmccurdy** [comment]
@zeux glTF Transform does quantize the vertex positions to 14 bits as well. I think the difference is that glTF Transform offers two quantization grids:

```
    --quantization-volume <volume>       Bounds for quantization grid.
                                         one of "mesh","scene", default: "mesh"
```

While 'mesh' may occasionally introduce seams, I've found it mostly fine as a default. Comparing the two modes in glTF Transform:


| mesh | scene |
|---|---|
| <img width="489" alt="Screenshot 2023-10-27 at 11 12 17 PM" src="https://github.com/zeux/meshoptimizer/assets/1848368/7c066f53-d991-4254-8a07-b8179c2a4590"> | <img width="529" alt="Screenshot 2023-10-27 at 11 10 12 PM" src="https://github.com/zeux/meshoptimizer/assets/1848368/a9b342e1-581b-458f-9dd5-fc6ac1774bac"> |

**Comment by zeux** [comment]
Yeah, seams are one reason why I've avoided per-mesh quantization. There's also issues with volume thickness (... this is really a spec issue in hindsight but oh well) in that if two meshes use the same material that has the extension, they need to use the same quantization scale. This also applies to skinning - since skinning matrices need to be adjusted, using a single skin for two meshes with different grids must also tie them together.

This is all doable but cumbersome which is part of why I've been avoiding fixing this :)

**Comment by zeux** [comment]
> the one thing that gltfpack could do is compute two quantization grids, one for skinned meshes and one for non-skinned meshes

Upon closer inspection this doesn't solve this scene fully, as it has 2 unskinned meshes in addition to skinned meshes that have scale 100; they are smaller and harder to see but it would still be an issue.

So besides using `-vpf` the only way to solve this appears to be per-mesh quantization scale which requires a fairly large change and some consideration as to what the defaults should be, which would need a lot of testing... I think this will not happen for the upcoming version as it's too significant of a change for a fairly core part of gltfpack, so I'm removing this from consideration for 0.20.
